File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int length(ListNode* head){
        ListNode* p = head;
        int l = 0;
        while(p != NULL){
            l++;
            p = p->next;
        }
        return l;
    }
    ListNode* solve(ListNode* head,int k){
        if(head == NULL)return NULL;
        if(k==0)return head->next;
        head->next = solve(head->next,k-1);
        return head;
    }
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int l = length(head);
        return solve(head,l-n);
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode h1 = head;
        ListNode h2 = head;

        for(int i = 0; i<n; i++) {
            if(h2 != null) {
                h2 = h2.next;
            } else {
                break;
            }
        }

        if(h2 == null) {
            return h1.next;
        }

        while(h2.next != null) {
            h1 = h1.next;
            h2 = h2.next;
        }

        ListNode temp = h1.next.next;
        h1.next = temp;
        return head;
    }
}